### PR TITLE
Add Command Binding sample apps for .NET 7 Command Binding blog post.

### PR DESCRIPTION
**Pull request by KlausLoeffelmann**
This PRs adds the sample solution for the Command Binding blogpost:

* Contains the Android .NET MAUI App.
* Contains the WinForms App which demos Command Binding in .NET 7.
* Contains the ViewModel, controlling both apps.
* Contains a demo project showing how to Unit Test a simple ViewModel.

MauiWinFormsEditor

### CODE REVIEW

File: Samples/CommandBinding/MauiEdit.UnitTest/UiControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        Assert.NotNull(dialogService);
+
+        MainFormController mainFormController = new(service)
+        {
+            // We clear the main document.
+            TextDocument = string.Empty
+        };
+
+        // We assert, that the New command is disabled, and cannot be called.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.UnitTest/UiControllerTest.cs:22]: Are these your comments or part of the template? Should avoid "We" and there is an unnecessary comma.

File: Samples/CommandBinding/MauiEdit.ViewModels/Base/ModalViewController.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    ///  generation.
+    /// </remarks>
+    [RelayCommand()]
+    protected virtual void OnOK()
+    {
+        Closed?.Invoke(this, new ModalViewResultEventArgs("OK"));
+    }
+
+    /// <summary>
+    /// Handler for the Cancel command. Bind this resulting command to the Cancel Button of a modal view.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Base/ModalViewController.cs:32]: nit: indent consistency

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 3]
```diff
@@ -0,0 +1,237 @@
+﻿using System.Diagnostics;
+using CommunityToolkit.Mvvm.ComponentModel;
+
+namespace MauiEdit.ViewModels;
+
+/// <summary>
+///  Main ViewModel for the WinForms/MAUI-Editor sample app. Encapsulated the full logic of the app.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:7]:
```suggestion
///  Main ViewModel for the WinForms/MAUI-Editor sample app. Encapsulates the full logic of the app.
```

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+public partial class MainFormController : ViewController
+{
+    // Holds the whole editor document.
+    private string? _textDocument;
+
+    // Backing field to iterate line-wise through the whole document.
+    private ReadOnlyMemory<char>[] _lines = Array.Empty<ReadOnlyMemory<char>>();
+
+    /// <summary>
+    ///  Gets the selection start- and end line based on SelectionIndex and SelectionLength.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:18]:
```suggestion
    ///  Gets the selection start and end line based on SelectionIndex and SelectionLength.
```

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    /// <summary>
+    ///  Provides the Document for the Editor.
+    /// </summary>
+    public string? TextDocument
+    {
+        get => _textDocument;
+
+        set
+        {
+            var wasEmpty = string.IsNullOrEmpty(_textDocument);
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:85]:
```suggestion
            bool wasEmpty = string.IsNullOrEmpty(_textDocument);
```

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+            if (SetProperty(ref _textDocument, value))
+            {
+                _lines = GetLinesFromDocument();
+
+                if (wasEmpty ^ string.IsNullOrEmpty(_textDocument))
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:91]: This is hard to follow. Why exclusive or?

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            // We shouldn't have this.
+            return "\n\r";
+        }
+        else if (TextDocument?.IndexOf("\r") > -1)
+        {
+            // WinUI TextBox.
+            return "\r";
+        }
+
+        // We don't have CrLf yet.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:164]: nit: consistency, have the comment in the block

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    // through a platform-depending binding.
+    // But this here works too.
+    private string? GetCarriageReturnStringBasedOnDocument()
+    {
+        if (TextDocument is null)
+        {
+            return null;
+        }
+
+        if (TextDocument?.IndexOf("\r\n") > -1)
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:148]: Presumably this would fail for "\n\r\n\r". Wouldn't it be better to look for just "\r" and check before and after?

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController_Commands.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+                i++;
+            }
+
+            if (CarriageReturn != "\r")
+            {
+                wrappedPart = wrappedPart.Replace("\r", CarriageReturn);
+            }
+
+            return (firstPart + wrappedPart.ToString() + CarriageReturn + lastPart);
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController_Commands.cs:138]: nit: use an interpolated string

File: Samples/CommandBinding/MauiEdit.WinForms/Program.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    ///  The main entry point for the application.
+    /// </summary>
+    [STAThread]
+    static void Main()
+    {
+        Application.SetHighDpiMode(HighDpiMode.SystemAware);
+        Application.EnableVisualStyles();
+        Application.SetCompatibleTextRenderingDefault(false);
+
+        // // Comment this out, to enable the simple WinForms binding demo.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.WinForms/Program.cs:19]: It is commented out

File: Samples/CommandBinding/MauiEdit.WinForms/Program.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        // Application.Run(new SimpleTestForm());
+        // return;
+        
+        var mainForm = new MainForm();
+
+        // We need a marshalling control.
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.WinForms/Program.cs:25]: Why do we need a marshalling control?

File: Samples/CommandBinding/MauiEdit.WinForms/UIControls/EditorControl.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+    
+    /// <summary>
+    ///  Raises the CursorPositionChanged event and makes CursorPosition bindable.
+    /// </summary>
+    /// <param name="e"></param>
+    protected virtual void OnCursorPositionChanged(EventArgs e)
+        => CursorPositionChanged?.Invoke(this, e);
+
+    // We want to listen to the EM_LINEFORMCHAR message. When it occurs, we know, the Selection has been
```
Comment (JeremyKuhne) [inline Samples/CommandBinding/MauiEdit.WinForms/UIControls/EditorControl.cs:80]:
```suggestion
    // We want to listen to the EM_LINEFORMCHAR message. When it occurs, we know the Selection has been
```

File: Samples/CommandBinding/MauiEdit.Maui/AppShell.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,9 @@
+﻿namespace MauiEdit;
+
+public partial class AppShell : Shell
+{
+	public AppShell()
+	{
+		InitializeComponent();
+	}
```
Comment (RussKie) [inline Samples/CommandBinding/MauiEdit.Maui/AppShell.xaml.cs:8]: Inconsistent use of spaces and tabs throughout.

File: Samples/CommandBinding/MauiEdit.WinForms/MauiEdit.WinForms.csproj [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+  <PropertyGroup>
+	<EnablePreviewFeatures>true</EnablePreviewFeatures>
+	<OutputType>WinExe</OutputType>
+    <TargetFramework>net7.0-windows</TargetFramework>
+    <UseWindowsForms>true</UseWindowsForms>
+    <LangVersion>10.0</LangVersion>
+	<Nullable>enable</Nullable>
+	<ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
```
Comment (RussKie) [inline Samples/CommandBinding/MauiEdit.WinForms/MauiEdit.WinForms.csproj:11]: Common properties are repeated many times across the solution. Please use Directory.Build.props/targets for those.
Also, consider using the [Central Package Management](https://learn.microsoft.com/nuget/consume-packages/Central-Package-Management).

Our samples should follow the best practices and guidelines, so that our samples help our consumers to learn and embrace those too.

File: Samples/CommandBinding/MauiEdit.ViewModels/Service/ServiceProviderExtension.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    static public T GetRequiredService<T>(this IServiceProvider serviceProvider)
+    {
+        if (serviceProvider.GetService(typeof(T)) is T service)
+        {
+            return service;
+        }
+        else
+        {
+            throw new InvalidOperationException($"Service {typeof(T).Name} not found.");
+        }
```
Comment (RussKie) [inline Samples/CommandBinding/MauiEdit.ViewModels/Service/ServiceProviderExtension.cs:14]: [nit] `else` is redundant.

File: Samples/CommandBinding/MauiEdit.WinForms/Program.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        WinFormsDialogService.RegisterWinFormsAsyncHelper(WinFormsAsyncHelper.GetInstance(mainForm));
+
+        var serviceProvider=SimpleServiceProvider.GetInstance();
```
Comment (RussKie) [inline Samples/CommandBinding/MauiEdit.WinForms/Program.cs:28]: [nit] Incorrect spacing.

File: Samples/CommandBinding/MauiEdit.UnitTest/UiControllerTest.cs (hunk 1 above)
Comment (KlausLoeffelmann) [reply Samples/CommandBinding/MauiEdit.UnitTest/UiControllerTest.cs:22]: Mine. I'll correct that.

File: Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs (hunk 6 above)
Comment (KlausLoeffelmann) [reply Samples/CommandBinding/MauiEdit.ViewModels/Controller/MainFormController.cs:91]: Every time the status changed to the exact opposite (The document _was_ empty, but now it's not anymore, or it was _not_ empty, but now it is), we need to inform the respective commands, that they should re-query their execution status. That controls the enable state of the respective menu items.

File: Samples/CommandBinding/MauiEdit.WinForms/Program.cs (hunk 11 above)
Comment (KlausLoeffelmann) [reply Samples/CommandBinding/MauiEdit.WinForms/Program.cs:25]: Simplicity, to be honest. To steer around doing the following with the `WindowsFormsSynchronizationContext`:

```
   public async Task<T?> InvokeAsync<T>(Func<T> function)
    {
        TaskCompletionSource<bool> taskCompletionSource = new();
        IAsyncResult? asyncResult = _marshallingControl.BeginInvoke(function);

        var registeredWaitHandle = ThreadPool.RegisterWaitForSingleObject(
               asyncResult.AsyncWaitHandle,
               new WaitOrTimerCallback(InvokeAsyncCallBack),
               taskCompletionSource,
               millisecondsTimeOutInterval: -1,
               executeOnlyOnce: true);

        await taskCompletionSource.Task;

        object? returnObject = _marshallingControl.EndInvoke(asyncResult);
        return (T?)returnObject;
    }
    ```

**Review by JeremyKuhne — APPROVED** [review]
